### PR TITLE
Send minimal lists of revisions from doc GET, _bulk_get, _revs_diff

### LINKED ISSUE #1752: Major inefficiencies replicating docs with many revisions
Replicating docs with large numbers of revisions (even without conflicts) becomes very inefficient, as SG and CBL end up sending large numbers of revIDs to each other. 

This is made worse by SG's very large default of 1000 for maxRevTreeDepth — 1000 revision IDs as a JSON array occupy about 39k bytes! (gzip cuts that down to about 22k, but that's still a lot.)
### Push
- [x] SG's response to `_revs_diff` includes a `possible_ancestors` property that lists revisions known to the server, so that the client can trim the history it sends when it then uploads the doc. However, SG includes _every_ known revision of the document. (Ironically this is much more bandwidth than if it skipped this optimization, since CBL would only send 20 revs max.) 

Solution: I think it should only send the leaf revision(s).
### Pull
- [x] When downloading a doc (or multiple docs with `_bulk_get`) the client uses `?revs=true` to request the revision's history, which the server adds as a `_revisions` property. Unfortunately this is the _entire_ history known to the server, i.e. up to 1000 revIDs. But the client usually only needs to know about a small number of revisions that are new to it.

Solution: The CouchDB protocol doesn't provide a way to limit this history, so we'd have to invent something comparable to what `_revs_diff` does.
- Option 1: Define a `revs_from` parameter whose value is an escaped JSON array of revIDs known to the client.
- Option 2: Use the existing `atts_since` parameter that's used to avoid sending unchanged attachments. This is also a list of revIDs known to the client. SG can safely use this to trim the history it returns. (Currently CBL only sends this if the document's current revision has attachments, and it sends all known revisions, up to 20. We'd need to change it to always send this parameter, but only include the leaf revisions.)
- [x] The above optimization won't help when a client pulls a document for the first time, since it doesn't know any of its revisions already. But it doesn't need 1000 revIDs, since it's only going to store the latest 20 (or whatever custom `maxRevTreeDepth` has been configured.)

Solution: Define a `revs_limit` parameter that specifies the maximum length of the history to send.

### PR DESCRIPTION
**Pull request by snej**
The REST API currently includes the revision's entire history in the
_revisions property of a document and the possible_ancestors property of
a _revs_diff response. This is vastly more than the client needs.
Sending a minimal number of revisions will decrease the bandwidth used
by replication, and save CPU cycles on both server and client.

There are a few parameter additions to the REST API — see #1752 
for details.

Docs:
https://github.com/couchbase/sync_gateway/wiki/Sending-Minimal-Histories

Fixes #1752

### CODE REVIEW

**Comment by adamcfraser** [comment]
The failure on the CI build looks like a known intermittent failure (https://github.com/couchbase/sync_gateway/issues/1736), unrelated to this PR.

File: db/revtree.go [hunk 1]
```diff
@@ -381,6 +377,20 @@ func ParseRevisions(body Body) []string {
...
 }
 
+// Splits out the "start" and "ids" properties from encoded revision list
+func splitRevisionList(revisions Body) (int, []string) {
+	start, ok := base.ToInt64(revisions["start"])
+	digests, _ := GetStringArrayProperty(revisions, "ids")
+	if ok && len(digests) > 0 && int(start) >= len(digests) {
+		return int(start), digests
+	} else {
+		base.TEMP("Couldn't split %+v -- start=%d, digests=%+v", revisions, start, digests)
```
Comment (adamcfraser) [inline db/revtree.go]: Remove TEMP logging, or switch to Warn?

File: db/revtree.go [hunk 2]
```diff
@@ -381,6 +377,20 @@ func ParseRevisions(body Body) []string {
 	return result
 }
 
+// Splits out the "start" and "ids" properties from encoded revision list
+func splitRevisionList(revisions Body) (int, []string) {
+	start, ok := base.ToInt64(revisions["start"])
+	digests, _ := GetStringArrayProperty(revisions, "ids")
```
Comment (adamcfraser) [inline db/revtree.go:383]: Do we need to worry about getting an error back from GetStringArrayProperty here, or should callers know they have a well-formed revisions Body by the time they call this?

File: db/revtree.go [hunk 3]
```diff
@@ -395,3 +405,25 @@ func encodeRevisions(revs []string) Body {
...
 	return Body{"start": start, "ids": ids}
 }
+
+// Given a revision history encoded by encodeRevisions() and a list of possible ancestor revIDs,
+// trim the history to stop at the first ancestor revID. If no ancestors are found, trim to
+// length maxUnmatchedLen.
+func trimEncodedRevisionsToAncestor(revs Body, ancestors []string, maxUnmatchedLen int) bool {
+	start, digests := splitRevisionList(revs)
+	if digests == nil {
+		return false
```
Comment (adamcfraser) [inline db/revtree.go:414]: What's the bool return value signify?  I thought at first it was "nothing trimmed", but I guess it's more like "there's nothing to evaluate for trimming"?  The current caller is ignoring the returned value, and the unit tests don't cover the case where it returns false.  I'm not sure whether it should just get removed, left as-is, or replaced with error.

File: rest/bulk_api.go [hunk 4]
```diff
@@ -295,8 +285,11 @@ func (h *handler) handleDumpChannel() error {
 // 	 ]
 // }
 func (h *handler) handleBulkGet() error {
-	includeRevs := h.getBoolQuery("revs")
 	includeAttachments := h.getBoolQuery("attachments")
+	revsLimit := 0
+	if h.getBoolQuery("revs") {
+		revsLimit = int(h.getIntQuery("revs_limit", math.MaxInt32))
```
Comment (adamcfraser) [inline rest/bulk_api.go:291]: The type conversion is a bit hard to follow here - passing a MaxInt32 to a method expecting uint64, then casting a returned uint64 back to int.  I think it should all build ok (even for a 32-bit build), but I'm wondering whether revsLimit could just be managed as an unsigned int, since negative values aren't valid anyway (I don't think?), to make it easier to follow.

File: rest/doc_api.go [hunk 5]
```diff
@@ -24,26 +25,42 @@ import (
...
-	if h.getBoolQuery("attachments") {
-		atts := h.getQuery("atts_since")
-		if atts != "" {
-			err := json.Unmarshal([]byte(atts), &attachmentsSince)
-			if err != nil {
-				return base.HTTPErrorf(http.StatusBadRequest, "bad atts_since")
+	// Check whether the caller wants a revision history, or attachment bodies, or both:
+	var revsLimit = 0
+	var revsFrom, attachmentsSince []string
+	{
```
Comment (adamcfraser) [inline rest/doc_api.go:33]: I haven't seen this kind of use of braces to create a statement without an operator/function before.  Are you doing something with the scoping of the variables that requires it (I couldn't spot anything), or is it just for functional grouping?

**Comment by adamcfraser** [comment]
Generally looks good - have a few minor comments/questions.

File: db/revtree.go (hunk 1 above)
Comment (snej) [inline db/revtree.go]: Oops, I meant to remove that line.

File: db/revtree.go (hunk 2 above)
Comment (snej) [inline db/revtree.go:383]: It's going to detect an error because `digests` will be nil. The details don't matter at this level because the functions that call this aren't returning specific errors, just success/failure. (If someone's screwed up this data they'll end up getting a 400 status back.)

File: db/revtree.go (hunk 3 above)
Comment (snej) [inline db/revtree.go:414]: It returns false if it fails, i.e. there's an error unpacking the revision list from `revs`.

Since this whole thing is just to optimize the size of the response, it didn't seem worthwhile to make a failure here cause the whole request to fail, which is why the caller ignores the return value.

File: rest/bulk_api.go (hunk 4 above)
Comment (snej) [inline rest/bulk_api.go:291]: The problem is that the default for `revs_limit` is "infinity", i.e. a huge value, but there's no `MaxInt` constant, as far as I know. So I had to choose `MaxInt32` because `MaxInt64` would break if assigned to an `int` in a 32-bit build.

The reason for having to cast back to `int` is that `getIntQuery` returns an `int64`, because there are other places that need to get 64-bit values from URL query params.

File: rest/doc_api.go (hunk 5 above)
Comment (snej) [inline rest/doc_api.go:33]: Mostly just for grouping, and to clarify the limited scope of the temporary variables declared inside.
